### PR TITLE
ECK Enterprise Search Helm Chart.

### PR DESCRIPTION
**Pull request by naemono**
## What is this change?

This adds an ECK EnterpriseSearch Helm Chart for deploying Enterprise Search instances using the ECK Operator.

## Testing

Installing with stack version `8.13.2` seems to work without issues (snapshot versions aren't quite working)
```
❯ helm upgrade -i eck-stack . -n elastic --create-namespace -f examples/enterprise-search/with-custom-configuration.yaml --set=eck-elasticsearch.version=8.13.2 --set=eck-kibana.version=8.13.2 --set=eck-enterprise-search.version=8.13.2
false
Release "eck-stack" has been upgraded. Happy Helming!
NAME: eck-stack
LAST DEPLOYED: Wed Apr 24 13:55:06 2024
NAMESPACE: elastic
STATUS: deployed
REVISION: 4
TEST SUITE: None
NOTES:
Elasticsearch ECK-Stack 0.11.0-SNAPSHOT has been deployed successfully!

To see status of all resources, run

kubectl get elastic -n elastic -l "app.kubernetes.io/instance"=eck-stack

More information on the Elastic ECK Operator, and its Helm chart can be found
within our documentation.

https://www.elastic.co/guide/en/cloud-on-k8s/current/index.html
```

Everything becomes healthy:
```
❯ kgp -n elastic
NAME                                    READY   STATUS    RESTARTS   AGE
elasticsearch-es-default-0              1/1     Running   0          3h40m
elasticsearch-es-default-1              1/1     Running   0          3h40m
elasticsearch-es-default-2              1/1     Running   0          3h40m
enterprise-search-ent-fc96ddf54-6dbks   1/1     Running   0          3h30m
kibana-kb-854b7dc9b4-fvlzb              1/1     Running   0          81s
```

All components green
```
❯ kubectl get elastic -n elastic -l "app.kubernetes.io/instance"=eck-stack
NAME                                                       HEALTH   NODES   VERSION   PHASE   AGE
elasticsearch.elasticsearch.k8s.elastic.co/elasticsearch   green    3       8.13.2    Ready   3h40m

NAME                                                                 HEALTH   NODES   VERSION   AGE
enterprisesearch.enterprisesearch.k8s.elastic.co/enterprise-search   green    1       8.13.2    3h40m

NAME                                  HEALTH   NODES   VERSION   AGE
kibana.kibana.k8s.elastic.co/kibana   green    1       8.13.2    3h40m
```

Required `elasticsearchRef` or `configRef` works: (i tried all variations of this from values commented out, to just being empty for one/both)
```
❯ helm template ent-search . -s templates/enterprisesearch.yaml
Error: execution error at (eck-enterprise-search/templates/enterprisesearch.yaml:22:5): At least one of configRef or elasticsearchRef is required
```

Enterprise Search integration with Kibana seems to work:
image

TODO:
- [x] additional verification/testing.
- [x] Documentation

### CODE REVIEW

File: deploy/eck-stack/charts/eck-enterprise-search/Chart.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+apiVersion: v2
+name: eck-enterprise-search
+description: Elastic Enterprise Search managed by the ECK operator
+kubeVersion: ">= 1.21.0-0"
+type: application
+version: 0.11.0-SNAPSHOT
+sources:
+  - https://github.com/elastic/cloud-on-k8s
+  - https://github.com/elastic/ent-search
```
Comment (pebrc) [inline deploy/eck-stack/charts/eck-enterprise-search/Chart.yaml:9]: Private repo. Let's leave that out.

File: deploy/eck-stack/examples/enterprise-search/with-custom-configuration.yaml [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    name: elasticsearch
+
+  spec:
+    enterpriseSearchRef:
+      name: enterprise-search
+
+eck-enterprise-search:
+  enabled: true
+
+  # Name of the Kibana instance.
```
Comment (pebrc) [inline deploy/eck-stack/examples/enterprise-search/with-custom-configuration.yaml:37]:
```suggestion
  # Name of the Enterprise Search instance.
```

File: deploy/eck-stack/charts/eck-enterprise-search/templates/enterprisesearch.yaml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+  annotations:
+    eck.k8s.elastic.co/license: basic
+    {{- with .Values.annotations }}
+    {{- toYaml . | nindent 4 }}
+    {{- end }}
+spec:
+  version: {{ required "An Enterprise Search version is required" .Values.version }}
+  count: {{ required "A pod count is required" .Values.count }}
+
+  # This is complicated, but seems required to catch both the situations where the key does not exist (commented out), and the key exists but is an empty map.
```
Comment (pebrc) [inline deploy/eck-stack/charts/eck-enterprise-search/templates/enterprisesearch.yaml:20]: This comment is redered e.g. when using `helm template` no big deal but it is not very useful in the rendered result. 
```suggestion
  {{- /*
  This is complicated, but seems required to catch both the situations where the key does not exist (commented out), and the key exists but is an empty map.
  */}} 
```

File: deploy/eck-stack/charts/eck-enterprise-search/examples/with-custom-configuration.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+config:
+  # define the exposed URL at which users will reach Enterprise Search
+  ent_search.external_url: https://my-custom-domain:3002
+  # define the exposed URL at which users will reach Kibana
+  kibana.host: https://kibana.my-custom-domain:5601
```
Comment (pebrc) [inline deploy/eck-stack/charts/eck-enterprise-search/examples/with-custom-configuration.yaml:5]: Maybe worth having these two commented out in the main `values.yaml` given how important they are. With a link to https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-enterprise-search-configuration.html maybe.

**Review by pebrc — APPROVED** [review]
LGTM!

File: deploy/eck-stack/charts/eck-enterprise-search/templates/enterprisesearch.yaml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  {{- if and (or (and (hasKey .Values "configRef") (eq 0 (len .Values.configRef))) (not (hasKey .Values "configRef"))) (or (and (hasKey .Values "elasticsearchRef") (eq 0 (len .Values.elasticsearchRef))) (not (hasKey .Values "elasticsearchRef"))) }}
+  {{ fail "At least one of configRef or elasticsearchRef is required" }}
```
Comment (thbkrkr) [inline deploy/eck-stack/charts/eck-enterprise-search/templates/enterprisesearch.yaml:22]: Wouldn't this deserve unit testing?

File: deploy/eck-stack/charts/eck-enterprise-search/values.yaml [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+# revisionHistoryLimit: 2
+
+# If you would prefer your sensitive data to be stored in a Secret, you can specify the name of the Secret reference.
+# In addition, if you do not want to use the `elasticsearchRef` mechanism or if you want to connect to an Elasticsearch
+# cluster not managed by ECK, you can manually configure Enterprise Search to access any available Elasticsearch cluster:
+# ref: https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-enterprise-search-configuration.html#k8s-enterprise-search-secret-configuration
+# ref: https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-enterprise-search-configuration.html#k8s-enterprise-search-connect-non-eck-es
+#
+configRef: {}
+  # secretName: elasticsearch-credentials
```
Comment (thbkrkr) [inline deploy/eck-stack/charts/eck-enterprise-search/values.yaml:89]:
```suggestion
  # secretName: enterprise-search-config
```

File: deploy/eck-stack/examples/enterprise-search/basic.yaml [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    name: elasticsearch
+
+  spec:
+    enterpriseSearchRef:
+      name: enterprise-search
+
+eck-enterprise-search:
+  enabled: true
+
+  # Name of the Kibana instance.
```
Comment (thbkrkr) [inline deploy/eck-stack/examples/enterprise-search/basic.yaml:37]:
```suggestion
  # Name of the Enterprise Search instance.
```

**Review by thbkrkr — APPROVED** [review]

File: deploy/eck-stack/charts/eck-enterprise-search/Chart.yaml (hunk 1 above)
Comment (naemono) [reply deploy/eck-stack/charts/eck-enterprise-search/Chart.yaml]: I honestly didn't know/see that. removed.

File: deploy/eck-stack/charts/eck-enterprise-search/templates/enterprisesearch.yaml (hunk 5 above)
Comment (naemono) [reply deploy/eck-stack/charts/eck-enterprise-search/templates/enterprisesearch.yaml:24]: I've added this. Honestly my head told me that you couldn't test failures, but clearly you can, as we do in the ECK-Elasticsearch chart as well. 🤷

File: deploy/eck-stack/charts/eck-enterprise-search/values.yaml (hunk 6 above)
Comment (naemono) [reply deploy/eck-stack/charts/eck-enterprise-search/values.yaml]: well, I was trying to convey that you could have an external es instance configuration in here as well, but I guess the comment beforehand notes that sufficiently.

**Pull request merged**